daemon/graphdriver/lcow/lcow.go

```diff
@@ -270,111 +551,123 @@ func (d *Driver) Remove(id string) error {
-		//
-		// Sandboxes need hot-adding in the case that there is a single global utility VM
-		// This will change for multiple instances with the lifetime changes.
-		if isSandbox {
-			hotAddRequired = true
-		}
-		d.cache[id] = cacheType{
-			uvmPath:   fmt.Sprintf("/mnt/%s", id),
+		cacheEntry = &cacheItem{
 			refCount:  1,
 			isSandbox: isSandbox,
 			hostPath:  vhdFilename,
+			uvmPath:   fmt.Sprintf("/mnt/%s", id),
+			isMounted: false, // we defer this as an optimisation
 		}
+		d.cache[id] = cacheEntry
+		logrus.Debugf("%s: added cache entry %+v", title, cacheEntry)
 	} else {
 		// Increment the reference counter in the cache.
-		cacheEntry = d.cache[id]
-		cacheEntry.refCount++
-		d.cache[id] = cacheEntry
-	}
-
-	cacheEntry = d.cache[id]
-	logrus.Debugf("%s %s: isSandbox %t, refCount %d", title, id, cacheEntry.isSandbox, cacheEntry.refCount)
-	d.cacheMu.Unlock()
-
-	if hotAddRequired {
-		logrus.Debugf("%s %s: Hot-Adding %s", title, id, vhdFilename)
-		if err := d.config.HotAddVhd(vhdFilename, cacheEntry.uvmPath); err != nil {
-			return "", cacheType{}, 0, fmt.Errorf("%s hot add %s failed: %s", title, vhdFilename, err)
-		}
+		logrus.Debugf("%s: locking cache item for increment", title)
+		entry.Lock()
+		entry.refCount++
+		logrus.Debugf("%s: releasing cache item for increment", title)
+		entry.Unlock()
+		logrus.Debugf("%s: incremented refcount on cache entry %+v", title, cacheEntry)
 	}
+	logrus.Debugf("%s: releasing cacheMutex", title)
+	d.cacheMutex.Unlock()
 
 	logrus.Debugf("%s %s success. %s: %+v: size %d", title, id, d.dir(id), cacheEntry, vhdSize)
-	return d.dir(id), cacheEntry, vhdSize, nil
+	return d.dir(id), nil
 }
 
 // Put does the reverse of get. If there are no more references to
 // the layer, it unmounts it from the utility VM.
 func (d *Driver) Put(id string) error {
-	title := "lcowdriver: put"
-	logrus.Debugf("%s %s", title, id)
-
-	if err := d.startUvm(fmt.Sprintf("put %s", id)); err != nil {
-		return err
-	}
+	title := fmt.Sprintf("lcowdriver: put: %s", id)
 
-	d.cacheMu.Lock()
-	// Bad-news if unmounting something that isn't in the cache.
+	logrus.Debugf("%s: locking cacheMutex", title)
+	d.cacheMutex.Lock()
 	entry, ok := d.cache[id]
 	if !ok {
-		d.cacheMu.Unlock()
+		logrus.Debugf("%s: releasing cacheMutex on error path", title)
+		d.cacheMutex.Unlock()
 		return fmt.Errorf("%s possible ref-count error, or invalid id was passed to the graphdriver. Cannot handle id %s as it's not in the cache", title, id)
 	}
 
-	// Are we just decrementing the reference count
+	// Are we just decrementing the reference count?
+	logrus.Debugf("%s: locking cache item for possible decrement", title)
+	entry.Lock()
 	if entry.refCount > 1 {
 		entry.refCount--
-		d.cache[id] = entry
-		logrus.Debugf("%s %s: refCount decremented to %d", title, id, entry.refCount)
-		d.cacheMu.Unlock()
+		logrus.Debugf("%s: releasing cache item for decrement and early get-out as refCount is now %d", title, entry.refCount)
+		entry.Unlock()
```

Comment:
That's fair. Yes, the debugging stuff is deliberately paranoid right now. Once it's stable, I can make changes such as this.